loadbalancer-standalone/src/main/java/com/rtbhouse/grpc/loadbalancer/standalone/AvailableServerList.java

```diff
@@ -0,0 +1,225 @@
+package com.rtbhouse.grpc.loadbalancer.standalone;
+
+import static java.lang.Math.ceil;
+
+import com.google.protobuf.ByteString;
+import io.grpc.lb.v1.Server;
+import io.grpc.lb.v1.ServerList;
+import java.net.InetSocketAddress;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class AvailableServerList {
+  private static final int SERVER_LIST_LENGTH_RATIO = 3;
+  private static final long STARTING_STATE = 0L;
+
+  private final LoadBalanceService lbService;
+  /**
+   * We aggregate the servers timestamps per service. key: name of service value: list of servers
+   * providing such service
+   */
+  private final Map<String, Set<BackendServer>> serviceToServers = new ConcurrentHashMap<>();
+
+  /** Indicates state of serviceToServers. Every add/remove operation should increase state. */
+  private final AtomicLong state = new AtomicLong(STARTING_STATE);
+
+  public AvailableServerList(LoadBalanceService lbService) {
+    this(lbService, 1000, 1200);
+  }
+
+  public AvailableServerList(
+      LoadBalanceService lbService, int timeBetweenUpToDateChecks, int timeToEvict) {
+    this.lbService = lbService;
+    runServersEvictor(timeBetweenUpToDateChecks, timeToEvict);
+  }
+
+  /**
+   * Runs an executor that periodically evicts outdated servers .
+   *
+   * @param frequency delay between consecutive executions of the evicting function (in
+   *     milliseconds)
+   * @param timeToEvict time after a server becomes outdated (in milliseconds)
+   */
+  private void runServersEvictor(int frequency, int timeToEvict) {
+    Executors.newScheduledThreadPool(1)
+        .scheduleAtFixedRate(
+            () -> evictOutdatedServers(timeToEvict), 0, frequency, TimeUnit.MILLISECONDS);
+  }
+
+  private void evictOutdatedServers(long timeToEvict) {
+    for (Map.Entry<String, Set<BackendServer>> serviceToServersKV : serviceToServers.entrySet()) {
+      boolean outdated = false;
+      long currentTime = System.currentTimeMillis();
+
+      String service = serviceToServersKV.getKey();
+      Set<BackendServer> servers = serviceToServersKV.getValue();
+
+      for (BackendServer server : servers) {
+        if (currentTime - server.getTimestamp() > timeToEvict && servers.remove(server)) {
```

Comment:
Czy tu nie ma ewentualności takiej, że przez dłuższy czas nie przepropaguje się informacja o tym, że server zevictował? W tym if'ie serwer jest usuwany z listy, ustawiana jest flaga outdated, a następnie uruchamiana jest metoda, tryToUpdate, która jak nazwa sugeruje może się nie powieść? I wtedy przez dłuższy czas do następnego zevictowania nie mamy informacji o tym, że serwer jest już niedostępny?

EDIT - chyba rozumiem, chodzi o to, żeby dwa razy nie updateować lb tym samym stanem listy serwerów.

Reply:
Tak, to prawda, chodzi głównie o sytuację np. podczas startu loadbalancera, kiedy nagle przychodzi dużo zgłoszeń od serwerów.